### ISSUE TITLE
VK_KHR_deferred_host_operations needs to be implemented before adding Raytracing Support

### ISSUE THREAD

**Issue by AntarticCoder** [KhronosGroup/MoltenVK#1953]
In issue #427 there has been discussion about the implementation of `VK_KHR_ray_tracing_pipeline` or `VK_KHR_ray_query`. Both of these extensions need the `VK_KHR_acceleration_structure` extension to work. To implement acceleration structures, we need the `VK_KHR_deferred_host_operation`, which allows for heavy operations that can take longer to do to be deferred until later. So as a small step towards raytracing, deferred_host_operations could be implemented. 

This issue is dependent on issue #427

Edit: I'd like to try and implement this, could somebody point me to which files would need to be modified as I'm not quite sure where to start

**Comment by natevm** [comment]
I’m not that familiar with MoltenVK, but I’m guessing the best first steps here would be to add a `vkCreateDeferredOperationKHR`and a `vkDestroyDeferredOperationKHR` that allocate / free an opaque `VkDeferredOperationKHR` object.

**Comment by AntarticCoder** [comment]
Yeah that seems simple but looking through the source files, I'm not able to find a file to put `VkDeferredOperationKHR`, I know it would have to go in the `/GPUObjects` subdirectory though, but it seems unclear to me what files `VkDeferredOperationKHR` would go into.

**Comment by AntarticCoder** [comment]
Would an opaque handle of `VkDeferredOperationKHR` fit into `GPUObjects/MVKSync.h`?

**Comment by natevm** [comment]
Personally I think that’s a good fit. This is a synchronization mechanism for sure, just a bit unique in that it’s about host-side asynchronous operations. 

Worst case we can move it around later on in a pull request.

**Comment by AntarticCoder** [comment]
Thanks for the insigh, I plan to start a PR on Monday/Tuesday

**Comment by AntarticCoder** [comment]
Opened a pull request, please link the issue with it: #1954

**Comment by natevm** [comment]
IMO, a union seems like a decent way to do it. So long as there is a clear organization that can be extended.

**Comment by AntarticCoder** [comment]
That's how I plan to do it, everytime a new vulkan command is implemented requiring deferred operations, you add the operations functions pointer to the union and also add the function to an enum that will manage with variable in the union should be used.

**Comment by natevm** [comment]
Indirectly related to this issue, does Metal RT support host-side acceleration structure construction? If it doesn’t, then is it possible to still support vk_khr_acceleration_structure?

**Comment by cdavis5e** [comment]
> Indirectly related to this issue, does Metal RT support host-side acceleration structure construction? (iiuc, that’s the reason this is a dependency of that, right?)

Nope. It only has device-side builds.

> If it doesn’t, then is it possible to still support vk_khr_acceleration_structure?

Yes, because host-side builds are [an optional feature](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap45.html#features-accelerationStructureHostCommands).

**Comment by natevm** [comment]
Nice work!! Looking forward to the next issue / PR.

Acceleration structures will be a tougher to implement I’d imagine, but if helps at all, I have a Vulkan ray tracing framework where I could test certain features as they’re added.

**Comment by AntarticCoder** [comment]
Thanks I just opened an issue #1956
